File: android/app/src/main/java/com/albbiz/map/data/EventsRepository.kt
// Bismillah Hir Rahman Nir Raheem
package com.albbiz.map.data

import android.net.Uri
import android.util.Log
import com.google.firebase.storage.FirebaseStorage
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.tasks.await

class EventsRepository {
    private val firestoreService = FirestoreService()
    private val storage = FirebaseStorage.getInstance()

    fun getEvents(): Flow<List<Event>> {
        return firestoreService.getEvents()
            .catch { e ->
                Log.e("AlbBizMap", "EventsRepo: Error in flow", e)
                emit(emptyList())
            }
    }

    suspend fun addEvent(event: Event, imageUri: Uri?): Result<String> {
        return try {
            // Upload image if provided
            val imageUrl = if (imageUri != null) {
                val filename = "events/${event.id}/image.jpg"
                val ref = storage.reference.child(filename)
                ref.putFile(imageUri).await()
                ref.downloadUrl.await().toString()
            } else null

            val finalEvent = event.copy(imageUrl = imageUrl)
            firestoreService.addEvent(finalEvent)
        } catch (e: Exception) {
            Log.e("AlbBizMap", "EventsRepo: Error adding event", e)
            Result.failure(e)
        }
    }

    fun getEventsByOrganizer(organizerId: String): Flow<List<Event>> {
        return firestoreService.getEventsByOrganizer(organizerId)
            .catch { e ->
                Log.e("AlbBizMap", "EventsRepo: Error fetching organizer events", e)
                emit(emptyList())
            }
    }

    // Deletes the Firestore doc and, if the event had a photo, its Storage file too.
    // Storage delete failures are logged but don't block the Firestore delete from
    // succeeding — an orphaned image is a much smaller problem than a stuck event
    // the owner can no longer remove.
    suspend fun deleteEvent(event: Event): Result<Unit> {
        return try {
            if (!event.imageUrl.isNullOrBlank()) {
                try {
                    storage.getReferenceFromUrl(event.imageUrl).delete().await()
                } catch (e: Exception) {
                    Log.e("AlbBizMap", "EventsRepo: Error deleting event image", e)
                }
            }
            firestoreService.deleteEvent(event.id)
        } catch (e: Exception) {
            Log.e("AlbBizMap", "EventsRepo: Error deleting event", e)
            Result.failure(e)
        }
    }
}

File: ios/AlbBizMap_IOS/AlbBizMap/Services/EventsRepository.swift
// Bismillah Hir Rahman Nir Raheem
import Foundation
import Combine
import FirebaseFirestore
import FirebaseStorage

public class EventsRepository: ObservableObject {
    public static let shared = EventsRepository()

    @Published public var events: [Event] = []

    private let db = Firestore.firestore()
    private lazy var eventsRef = db.collection("events")
    private let storage = Storage.storage()

    private var eventsListener: ListenerRegistration?

    private init() {
        startListening()
    }

    private func startListening() {
        eventsListener = eventsRef
            .order(by: "date")
            .addSnapshotListener { [weak self] snapshot, error in
                guard let self else { return }
                if let error {
                    print("EventsRepository: error listening for events: \(error)")
                    return
                }
                let events: [Event] = snapshot?.documents.compactMap { doc in
                    Event.fromMap(id: doc.documentID, map: doc.data())
                } ?? []
                DispatchQueue.main.async {
                    self.events = events
                }
            }
    }

    // Uploads the event photo (if any) first, then writes the event with its
    // URL already set — mirrors Android's EventsRepository.addEvent.
    public func addEvent(_ event: Event, photoData: Data? = nil) async -> Result<String, Error> {
        let docRef = eventsRef.document()
        do {
            var imageUrl: String? = nil
            if let photoData {
                let filename = "events/\(docRef.documentID)/image.jpg"
                let ref = storage.reference().child(filename)
                let metadata = StorageMetadata()
                metadata.contentType = "image/jpeg"
                _ = try await ref.putDataAsync(photoData, metadata: metadata)
                imageUrl = try await ref.downloadURL().absoluteString
            }
            var finalEvent = event
            finalEvent.id = docRef.documentID
            finalEvent.imageUrl = imageUrl
            try await docRef.setData(finalEvent.toMap().compactMapValues { $0 })
            return .success(docRef.documentID)
        } catch {
            return .failure(error)
        }
    }

    // Events a given user submitted — powers "My Events" on the Profile
    // screen, mirroring FirestoreService's getBusinessesByOwner pattern.
    public func getEventsByOrganizer(organizerId: String) -> [Event] {
        events.filter { $0.organizerId == organizerId }
    }

    // Deletes the Firestore doc and, if the event had a photo, its Storage
    // file too. A Storage delete failure is logged but doesn't block the
    // Firestore delete from succeeding — an orphaned image is a much
    // smaller problem than a stuck event the owner can no longer remove.
    public func deleteEvent(_ event: Event) async -> Result<Void, Error> {
        do {
            if let imageUrl = event.imageUrl, !imageUrl.isEmpty {
                do {
                    try await storage.reference(forURL: imageUrl).delete()
                } catch {
                    print("EventsRepository: error deleting event image: \(error)")
                }
            }
            try await eventsRef.document(event.id).delete()
            return .success(())
        } catch {
            return .failure(error)
        }
    }
}
